src/porepy/models/solution_strategy.py

```diff
@@ -1026,6 +1018,245 @@ def darcy_flux_storage_keywords(self) -> list[str]:
+                previous nonlinear iteration.
+            residual: The current residual vector of the nonlinear system.
+            nl_params: Dictionary of nonlinear solver parameters. Is for now expected
+                to contain "nl_convergence_tol" (tolerance for the nonlinear increment
+                norm) and "nl_convergence_tol_res" (tolerance for the residual norm).
+
+        Returns:
+            A tuple with two elements. A boolean indicating whether the problem has
+            converged and a boolean indicating whether it has diverged.
+
+        """
+        if not self._is_nonlinear_problem:
+            # At least for the default direct solver, scipy.sparse.linalg.spsolve, no
+            # error (but a warning) is raised for singular matrices, but a nan solution
+            # is returned. We check for this.
+            diverged = bool(np.any(np.isnan(nonlinear_increment)))
+            converged: bool = not diverged
+            return converged, diverged
+
+        # First a simple check for nan values.
+        if np.any(np.isnan(nonlinear_increment)):
+            # If the solution contains nan values, we have diverged.
+            return False, True
+
+        # Array of all residual norms
+        residual_norms = self.compute_residual_norm(residual)
+
+        # Array of all variable increment norms
+        nonlinear_increment_norms = self.compute_nonlinear_increment_norm(
+            nonlinear_increment
+        )
+
+        # Reference values for increment norms:
+        # During the first iteration, initialize bookkeeping for reference norms
+        if self.nonlinear_solver_statistics.num_iteration == 1:
+            self.fixed_reference_nonlinear_increment_norms = {
+                key: False for key in nonlinear_increment_norms
+            }
+            self.reference_nonlinear_increment_norms = {
+                key: 1.0 for key in nonlinear_increment_norms
+            }
+
+        # If any of the increment norms does not have a reference value (i.e., at least
+        # one element in fixed_reference_nonlinear_increment_norms is False), use the
+        # first iterate solution as the reference.
+        if not all(self.fixed_reference_nonlinear_increment_norms.values()):
+            reference_solution = self.equation_system.get_variable_values(
+                iterate_index=0
+            )
+            reference_solution_norms = self.compute_nonlinear_increment_norm(
+                reference_solution
+            )
+        # Assign valid reference norms where still missing
+        for key, is_fixed in self.fixed_reference_nonlinear_increment_norms.items():
+            if not is_fixed:
+                _norm = reference_solution_norms[key]
+                # Reference norms should not be zero or NaN.
+                if not np.isclose(_norm, 0.0) and not np.isnan(_norm):
+                    self.reference_nonlinear_increment_norms[key] = _norm
```

Comment:
Please explain why we need both reference_nonlinear_increment_norms and fixed?reference_nonlinear_increment_norms.

Reply:
Dealing with reference values has been fully restructured.